staging/src/k8s.io/apiserver/pkg/storage/etcd3/paging_etcd_client.go

```diff
@@ -0,0 +1,110 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd3
+
+import (
+	"context"
+
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// apiserver only uses the following etcd v3 client options without any sort order and target
+// it default falls back to sorted by keys in lexiographically ascending order
+
+// key
+// WithRange
+// WithLimit
+// WithRev
+
+// etcd mvcc layer increment limit by one to fetch for 'more' flag
+
+// current behavior
+// 1. only set WithLimit when paging is enabled (by default yes) and limit > 0 (it can be 0 if client not set)
+//    1.1 if the resource version is provided, it set up the rev and returnedRV in legacy cases and RVMatchExact
+//    1.2
+// 2. if predicate continue token is provided, restore key and withRev from it
+
+// proposed behavior
+// 1. configure a maximum limit to etcd when limit is not provided by client and construct the full list result
+
+const defaultLimit = 500
+
+type etcdClient struct {
+	c            *clientv3.Client
+	paingEnabled bool
+	limit        int64
+}
+
+// Get implements clientv3.KV.Get
+// APIServer only use key, WithRange, WithLimit and WithRev options without any sort order and target
+// it default falls back to sorted by keys in lexiographically ascending order
+// GetResponse is a type alias of "go.etcd.io/etcd/api/v3/etcdserverpb".RangeResponse
+
+// It's better to add the paging support in APIServer because it knows calling etcd with ascending order and sorting target is key
+func (ec *etcdClient) Get(ctx context.Context, key string, limit int64, getOpts ...clientv3.OpOption) (*clientv3.GetResponse, error) {
+	if !ec.paingEnabled || (limit > 0 && limit <= ec.limit) {
+		return ec.c.Get(ctx, key, getOpts...)
+	}
+
+	// split unlimited list requests into multiple pages in case of cache miss to reduce etcd load
+	// this may increase round trip time (alternatively consider etcd server side range streaming)
+	originalLimit := int(limit)
+	limit = ec.limit
+
+	op := clientv3.OpGet(key, getOpts...)
+	rev := op.Rev() // if it is 0, stored it in the first query
+	// Potential WithPrefix will set the end finally and the key is not expected to be empty
+	end := string(op.RangeBytes())
+
+	result := &clientv3.GetResponse{}
+	result.Kvs = make([]*mvccpb.KeyValue, 0, originalLimit)
+
+	var lastKey []byte
+	for {
+		opts := []clientv3.OpOption{clientv3.WithRange(end), clientv3.WithRev(rev), clientv3.WithLimit(limit)}
+		rangeResp, err := ec.c.Get(ctx, key, opts...)
+		if err != nil {
+			return nil, err
+		}
+
+		// [Debatable] set result header and count in the first query
+		if result.Header == nil && result.Count == 0 {
+			result.Header = rangeResp.Header
+			result.Count = rangeResp.Count
+		}
+
+		// if no initial revision is provided, use the revision returned in the first query
+		if rev == 0 {
+			rev = rangeResp.Header.Revision
+		}
+
+		for _, kv := range rangeResp.Kvs {
```

Comment:
This is an inner copy loop, which I think is worth optimising further - once we are comfortable with the general approach.

(In particular: move the conditional out of the loop, and use a single `append(result.kvs, resp.kvs...)` rather than a loop with repeat single-element appends)